pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -204,3 +205,24 @@ func (r *Requirements) UnmarshalJSON(b []byte) error {
 	*r = NewRequirements(requirements...)
 	return nil
 }
+
+func (r *Requirements) String() string {
+	var sb strings.Builder
+	for key, req := range r.requirements {
+		var values []string
+		if !req.IsComplement() {
+			values = req.Values().List()
+		} else {
+			values = req.ComplementValues().List()
```

Comment:
The `Type()` method will return a NotIn or an Exists operator for complement sets.

```go
	tmp := v1alpha5.NewRequirements(
		v1.NodeSelectorRequirement{Key: "key-1", Operator: v1.NodeSelectorOpNotIn, Values: []string{"bar", "baz"}},
		v1.NodeSelectorRequirement{Key: "key-2", Operator: v1.NodeSelectorOpExists},
		v1.NodeSelectorRequirement{Key: "key-3", Operator: v1.NodeSelectorOpDoesNotExist},
		v1.NodeSelectorRequirement{Key: "key-4", Operator: v1.NodeSelectorOpIn, Values: []string{"a", "b", "c"}},
	)

	a := tmp.String()
// Output: key-4 In [a b c], key-1 NotIn [bar baz], key-2 Exists [], key-3 DoesNotExist []
```

The exists/does not exist show an empty set, but not sure if that's worth fixing since the intent is still clear.

Reply:
Beautiful